coderd/workspaces.go

```diff
@@ -1200,6 +1204,22 @@ func convertWorkspaceTTLMillis(i sql.NullInt64) *int64 {
 	return &millis
 }
 
+// Calculate the time of the upcoming working deletion, if applicable; otherwise, return an empty time.Time struct.
+// Workspaces may have impending deletions if InactivityTTL feature is turned on and the workspace is inactive.
+func calculateImpendingDeletion(workspace database.Workspace, template database.Template) time.Time {
+	var (
+		year, month, day = time.Now().Date()
+		beginningOfToday = time.Date(year, month, day, 0, 0, 0, 0, time.Now().Location())
+	)
+	// If InactivityTTL is turned off (set to 0), if the workspace has already been deleted,
+	// or if the workspace was used sometime within the last day, there is no impending deletion
+	if template.InactivityTTL == 0 || workspace.Deleted || workspace.LastUsedAt.After(beginningOfToday) {
+		return time.Time{}
+	}
+
+	return workspace.LastUsedAt.Add(time.Duration(template.InactivityTTL) * time.Nanosecond)
+}
```

Comment:
I tried to write some tests for this in `workspaces_test.go` but I was unsure how to artificially set `workspace.lastUsedAt`:

````
	t.Run("ImpendingDeletion", func(t *testing.T) {
		t.Parallel()
		client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
		user := coderdtest.CreateFirstUser(t, client)
		version := coderdtest.CreateTemplateVersion(t, client, user.OrganizationID, nil)
		coderdtest.AwaitTemplateVersionJob(t, client, version.ID)
		template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID, func(ctr *codersdk.CreateTemplateRequest) {
			ctr.InactivityTTLMillis = ptr.Ref(10 * 24 * time.Hour.Milliseconds()) // delete after 10 days of inactivity
		})

		workspace := coderdtest.CreateWorkspace(t, client, user.OrganizationID, template.ID, func(ctr *codersdk.CreateWorkspaceRequest) {

		}) // this is not inactive; last used sometime in the last day

		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
		defer cancel()

		workspaceNew, err := client.WorkspaceByOwnerAndName(ctx, codersdk.Me, workspace.Name, codersdk.WorkspaceOptions{})
		require.NoError(t, err)
		require.Equal(t, workspaceNew.ImpendingDeletion, time.Time{})
	})
````

Reply:
This is a lot, instead you should just test the helper function.

Make a new file called `workspaces_internal_test.go`. This allows you to call the unexported method for testing. Your file will look like this:

```go
package coderd

import (
	"github.com/coder/coder/coderd/database"
	"github.com/stretchr/testify/require"
	"testing"
	"time"
)

func Test_calculateImpendingDeletion(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name     string
		workspace database.Workspace
		template database.Template
		expected time.Time
	} {
		// Test cases here. Might want to make some helpers to create
		// workspace and templates with given field values.
	}

	for _, tc := range testCases {
		tc :=tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			found := calculateImpendingDeletion(tc.workspace, tc.template)
			require.WithinDuration(t, tc.expected, found, time.Second, "incorrect impending deletion"
		})
	}
}
```

Reply:
@Emyrk Added!